### PR TITLE
Enable Package Publishing and Generating Package Version from AssemblyInfo.cs

### PR DESCRIPTION
**Pull request by holycrepe**
**Major changes**: 

1. Use NuGet executable packaged with [VSTS](https://www.visualstudio.com/en-us/products/visual-studio-team-services-vs.aspx) credential provider 
2. Add `GenerateVersions` task to `EstablishNugetPaths` target, which will replace the `$version$` placeholder with the version from the first project that has a valid `Properties/AssemblyInfo.cs` file
3. Add `NuGetPush` task to `CreatePackage` target, which optionally publishes the package to a remote repository. 

**New Project Properties**:

- Remote repository access (`Repository`, `ApiKey`)
- Publishing settings (`PublishPackage`). 

    - `PublishPackage` can be `Always`, `Never`, or `Auto`. 
    - Currently, `Auto` will publish the package unless the `ConfigurationName` is `Debug`.

### CODE REVIEW

File: src/NuProj.ProjectSystem.12/NuProj.ProjectSystem.12.csproj [hunk 1]
```diff
@@ -80,6 +80,10 @@
     </Reference>
     <Reference Include="Microsoft.VisualStudio.Shell.Immutable.10.0" />
     <Reference Include="Microsoft.VisualStudio.Shell.Immutable.11.0" />
+    <Reference Include="Microsoft.VisualStudio.Validation, Version=14.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
+      <HintPath>..\packages\Microsoft.VisualStudio.Validation.14.1.111\lib\net45\Microsoft.VisualStudio.Validation.dll</HintPath>
```
Comment (AArnott) [inline src/NuProj.ProjectSystem.12/NuProj.ProjectSystem.12.csproj]: I don't think we should be using MS.VS.Validation 14.1 on VS 12.

File: src/NuProj.Packages/NuProj.nuproj [hunk 2]
```diff
@@ -51,7 +51,10 @@ There is also a Visual Studio extension which you find under http://bit.ly/NuPro
     <Content Include="$(BasePath)raw\Additional\Microsoft.Common.NuProj.targets">
       <Link>tools\Microsoft.Common.NuProj.targets</Link>
     </Content>
-    <Content Include="..\packages\NuGet.CommandLine.2.8.6\tools\NuGet.exe">
+    <Content Include="..\NuGet.VSTS\tools\CredentialProvider.VSS.exe">
```
Comment (AArnott) [inline src/NuProj.Packages/NuProj.nuproj]: Are these binaries available via NuGet so we don't have to check in binaries to git?

File: src/NuProj.ProjectSystem.12/NuProj.ProjectSystem.12.csproj (hunk 1 above)
Comment (holycrepe) [inline src/NuProj.ProjectSystem.12/NuProj.ProjectSystem.12.csproj]: Oh okay, I think this package may have been automatically updated via the VS package manager. I'll revert this change

File: src/NuProj.ProjectSystem.12/Rules/general.xaml [hunk 3]
```diff
@@ -1,33 +1,46 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (AArnott) [inline src/NuProj.ProjectSystem.12/Rules/general.xaml]: You've made a lot of whitespace changes to this file which makes it hard to filter out the signal from the noise. I'd prefer if you could adjust your pull request to only include meaningful changes, and save any whitespace changes you want to make for a separate PR so we can more efficiently review your changes.

File: src/NuProj.ProjectSystem.14/NuProj.ProjectSystem.14.csproj [hunk 4]
```diff
@@ -108,7 +108,7 @@
       <Private>True</Private>
     </Reference>
     <Reference Include="Microsoft.VisualStudio.Validation, Version=14.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
-      <HintPath>..\packages\Microsoft.VisualStudio.Validation.14.0.50702\lib\net45\Microsoft.VisualStudio.Validation.dll</HintPath>
+      <HintPath>..\packages\Microsoft.VisualStudio.Validation.14.1.111\lib\net45\Microsoft.VisualStudio.Validation.dll</HintPath>
```
Comment (AArnott) [inline src/NuProj.ProjectSystem.14/NuProj.ProjectSystem.14.csproj]: This makes NuProj require VS 2015 Update 1. What do we gain by requiring that?

File: src/NuProj.ProjectSystem.14/NuProj.ProjectSystem.14.csproj [hunk 5]
```diff
@@ -334,6 +334,9 @@
   <Import Project="..\packages\Microsoft.VisualStudio.ProjectSystem.SDK.Tools.14.0.50721-pre\build\Microsoft.VisualStudio.ProjectSystem.Sdk.Tools.targets" Condition="Exists('..\packages\Microsoft.VisualStudio.ProjectSystem.SDK.Tools.14.0.50721-pre\build\Microsoft.VisualStudio.ProjectSystem.Sdk.Tools.targets')" />
   <Import Project="..\packages\Microsoft.VisualStudio.SDK.VsixSuppression.14.0.50703\build\Microsoft.VisualStudio.SDK.VsixSuppression.targets" Condition="Exists('..\packages\Microsoft.VisualStudio.SDK.VsixSuppression.14.0.50703\build\Microsoft.VisualStudio.SDK.VsixSuppression.targets')" />
   <Import Project="..\packages\Nerdbank.GitVersioning.1.3.13\build\dotnet\Nerdbank.GitVersioning.targets" Condition="Exists('..\packages\Nerdbank.GitVersioning.1.3.13\build\dotnet\Nerdbank.GitVersioning.targets')" />
+  <PropertyGroup>
+    <PostBuildEvent>copy "$(TargetDir)*.vsix" "$(SolutionDir)..\..\..\*.vsix"</PostBuildEvent>
```
Comment (AArnott) [inline src/NuProj.ProjectSystem.14/NuProj.ProjectSystem.14.csproj]: Why do we want to do that?

File: src/NuProj.Packages/NuProj.nuproj (hunk 2 above)
Comment (holycrepe) [inline src/NuProj.Packages/NuProj.nuproj]: I could not find such a package; as far as I could tell, it is only available on a VSTS feed. 
We could proceed one of three ways:

1. I could manually publish a NuGet package with the credential provider
2. Revert the change and let users manually overwrite the files themselves, or
3. Keep the binaries as-is

Let me know which you think would be best

File: src/NuProj.ProjectSystem.14/NuProj.ProjectSystem.14.csproj (hunk 5 above)
Comment (holycrepe) [inline src/NuProj.ProjectSystem.14/NuProj.ProjectSystem.14.csproj]: Sorry this was for testing purposes. I'll revert this

File: src/NuProj.ProjectSystem.14/NuProj.ProjectSystem.14.csproj (hunk 4 above)
Comment (holycrepe) [inline src/NuProj.ProjectSystem.14/NuProj.ProjectSystem.14.csproj]: This might have also been done by the VS package manager. I'll revert this

File: src/NuProj.Targets/NuProj.props [hunk 6]
```diff
@@ -16,6 +16,7 @@
     <EmbedSourceFiles Condition=" '$(EmbedSourceFiles)' == '' ">False</EmbedSourceFiles>
     <RequireLicenseAcceptance Condition=" '$(RequireLicenseAcceptance)' == '' ">False</RequireLicenseAcceptance>
     <DevelopmentDependency Condition=" '$(DevelopmentDependency)' == '' ">False</DevelopmentDependency>
+    <PublishPackage Condition=" '$(PublishPackage)' == '' ">Auto</PublishPackage>
```
Comment (AArnott) [inline src/NuProj.Targets/NuProj.props]: No, I don't think we can accept this. Publishing a package should *not* be done as part of a Build. No way. 
As part of another MSBuild target (maybe Deploy) *perhaps*. But absolutely not can we have Build, which is generally a local machine operation, publish the result somewhere.

File: src/NuProj.ProjectSystem.12/Rules/general.xaml (hunk 3 above)
Comment (holycrepe) [inline src/NuProj.ProjectSystem.12/Rules/general.xaml]: Sorry for this - it seems I have an extension that automatically reformats xaml files on save. I will revert such files in the pull request and re-commit shortly

**Comment by AArnott** [comment]
I appreciate the time and care you've taken to prepare these changes, @holycrepe. I think the change where you read the version from the assembly may be good as many folks have asked for it. But we can't have nuproj publish during a build (release-only or not). And the whitespace changes should be in a separate PR.
Can you accommodate that?

**Comment by holycrepe** [comment]
Thanks for your review and comments. Should I adjust to integrate with a Deploy target, or just forget about it?

**Comment by holycrepe** [comment]
Yes I'll make a separate PR with version changes only. I'll make sure no whitespace changes occur

**Comment by holycrepe** [comment]
In my particular use case I have a Deploy build and have it integrated there, but I understand your reasoning behind not publishing a package during the Build process. We could still make use of the NuGetPush task by moving it into a separate MSBuild target (such as `Deploy`, like you suggested). Let me know if that would be useful

**Comment by holycrepe** [comment]
Created PR #203 for reading version from assembly.

Let me know how you'd like me to integrate `NuGetPush` task if you think that may be useful, and I'll make a separate PR
